docs: CSV design for operator cleanup (#2035)
operator-framework/operator-lifecycle-manager: doc/design/building-your-csv.md

                - list
    strategy: deployment
```

> ## Operator Cleanup
> The CSV can opt-into cleanup of the Custom Resources (CRs) reconciled by the operator by a setting on the spec which OLM 
> will honor. 

Reviewer: ```suggestion
## Operator Cleanup

The CSV can opt-into cleanup of the Custom Resources (CRs) reconciled by the operator by a setting on the spec which OLM will honor.
```

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
...
  spec:
  ...
    cleanup: 
      enabled: true
```

> When OLM sees that the CSV cleanup setting is set to `true` it will automatically put the `operatorframework.io/cleanup-apis`
> finalizer on the CSV. When the CSV is deleted by the user, kubernetes sets the deletion timestamp on the CSV but the finalizer prevents the 
> CSV from being deleted. OLM then finds and deletes all CRs managed by the operator by looking at the operator's `OwnedCRDs` section of the CSV
> and then finding these objects on-cluster. By deleting the CRs before removing the operator, OLM allows the operator to run custom cleanup logic
> on the CRs (via its own finalizers). Once all CRs have been deleted from the cluster, OLM removes the `operatorframework.io/cleanup-apis` finalizer 
> from the CSV which enables the operator to be deleted from the cluster. 

Reviewer: Just had a couple of minor nits. I tried to take a stab at that last sentence to make it clearer why Kubernetes was going to remove those resources after removing the finalizer.

```suggestion
When OLM sees that the CSV cleanup setting is set to `true`, it will automatically put the `operatorframework.io/cleanup-apis` finalizer on the CSV. When the CSV is deleted by the user, Kubernetes will set the deletion timestamp on the CSV, but the finalizer will prevent the CSV from being deleted. OLM will then find and delete all CRs managed by the operator by looking at the operator's `spec.OwnedCRDs` section of the CSV
and then finding these objects on-cluster. By deleting the CRs before removing the operator, OLM allows the operator to run custom cleanup logic on the CRs (via its own finalizers). Once all CRs have been deleted from the cluster, OLM then removes the `operatorframework.io/cleanup-apis` finalizer from the CSV, which signals to Kubernetes that the operator is ready to be deleted from the cluster.
```

> For more detailed information see the [OLM doc site](https://olm.operatorframework.io/docs/advanced-tasks/operator-cleanup/). 

Reviewer: ```suggestion
For more detailed information, see the [OLM doc site](https://olm.operatorframework.io/docs/advanced-tasks/operator-cleanup/).
```
